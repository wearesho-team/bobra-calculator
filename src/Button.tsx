import * as React from "react";
import { ControlContext, ControlContextValue } from "./ControlContext";

export interface ButtonProps extends React.HTMLProps<HTMLButtonElement> {
    increase: boolean,
}

export const Button: React.FC<ButtonProps> = (props) => {
    const { increase, children, ...childProps } = props;

    const context = React.useContext(ControlContext);
    const isDisabled = increase
        ? context.value >= context.max
        : context.value <= context.max;
    const handleClick = React.useCallback(() => {
        if (isDisabled) {
            return;
        }

        const delta = increase ? 1 : -1;
        context.onChange(context.value + delta * context.step);
    }, [ isDisabled, increase, context.value, context.step ]);

    return (
        <button {...props} disabled={isDisabled} onClick={handleClick}>
            {children}
        </button>
    );
};
