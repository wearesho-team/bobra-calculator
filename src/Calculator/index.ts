export * from "./Calculator";
export * from "./CalculatorContext";
export * from "./CalculatorProps";
