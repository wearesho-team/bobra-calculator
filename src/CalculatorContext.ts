import * as React from "react";

export interface CalculatorContextValue {
    amount: {
        min: number;
        max: number;
        value: number;
        step: number;
        onChange: (next: number) => number;
    },
    term: {
        min: number;
        max: number;
        value: number;
        step: number;
        onChange: (next: number) => number
    },
    interest: {
        rate: number,
        amount: number;
    };
}

export const CalculatorContextDefaultValue: CalculatorContextValue = {
    amount: {
        min: 250,
        max: 2000,
        value: 250,
        step: 50,
        onChange: (next) => next,
    },
    term: {
        min: 5,
        max: 30,
        value: 5,
        step: 1,
        onChange: (next) => next,
    },
    interest: {
        rate: 0.0175,
        amount: 0,
    },
};

export const CalculatorContext = React.createContext<CalculatorContextValue>(CalculatorContextDefaultValue);
