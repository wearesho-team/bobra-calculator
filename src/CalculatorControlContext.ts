import * as React from "react";

export interface CalculatorControlContextValue {
    min: number,
    max: number,
    step: number,
    value: number,
    onChange: (next: number) => number,
}

export const CalculatorControlContext = React.createContext<CalculatorControlContextValue>({
    min: 0,
    max: 0,
    step: 1,
    value: 0,
    onChange: (next) => next,
});
