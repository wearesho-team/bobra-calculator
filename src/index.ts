export * from "./Controller";
export * from "./Button";
export * from "./Context";
export * from "./ControlContext";
export * from "./ControlWrapper";
export * from "./DateLabel";
export * from "./Input";
export * from "./Label";
export * from "./SummaryButton";
export * from "./Range";
