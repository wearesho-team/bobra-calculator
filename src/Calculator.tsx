import * as React from "react";
import { CalculatorContext, CalculatorContextValue } from "./CalculatorContext";

export interface CalculatorState {
    term: number,
    amount: number,
}

export interface CalculatorProps {
    interestRate: number,
    amount: {
        max: number,
        min: number,
        initial?: number,
        step?: number,
    },
    term?: {
        max: number,
        min: number,
        initial?: number,
        step?: number,
    },

    onAmountChange?: (nextAmount: number) => void,
    onTermChange?: (nextTerm: number) => void,
}

export class Calculator extends React.PureComponent<CalculatorProps, CalculatorState> {
    public readonly state: CalculatorState = {
        term: this.props.term.initial || Math.round((this.props.term.min + this.props.term.max) / 2),
        amount: this.props.amount.initial || Math.round((this.props.amount.min + this.props.amount.max) / 2),
    };

    public get interest(): { amount: number, rate: number } {
        return {
            amount: Math.round(this.state.term * this.props.interestRate * this.state.amount),
            rate: this.props.interestRate,
        };
    }

    public render() {
        return <CalculatorContext.Provider value={this.contextValue} children={this.props.children}/>
    }

    protected get contextValue(): CalculatorContextValue {
        return {
            amount: {
                value: this.state.amount,
                min: this.props.amount.min,
                max: this.props.amount.max,
                step: this.props.amount.step,
                onChange: this.handleAmountChange,
            },
            term: {
                value: this.state.term,
                min: this.props.term.min,
                max: this.props.term.max,
                step: this.props.term.step,
                onChange: this.handleTermChange,
            },
            interest: this.interest,
        };
    }

    protected handleAmountChange = (nextAmount: number): number => {
        nextAmount = Math.max(this.props.amount.min, nextAmount);
        nextAmount = Math.min(this.props.amount.max, nextAmount);

        this.setState({ amount: nextAmount });

        return nextAmount;
    };

    protected handleTermChange = (nextTerm: number): number => {
        if (this.props.term === undefined) {
            return;
        }

        nextTerm = Math.max(this.props.term.min, nextTerm);
        nextTerm = Math.min(this.props.term.max, nextTerm);

        this.setState({ term: nextTerm });
        return nextTerm;
    };
}
