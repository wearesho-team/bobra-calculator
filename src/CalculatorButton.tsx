import * as React from "react";
import { CalculatorControlContext, CalculatorControlContextValue } from "./CalculatorControlContext";

export interface CalculatorButtonProps extends React.HTMLProps<HTMLButtonElement> {
    increase: boolean,
}

export class CalculatorButton extends React.PureComponent<CalculatorButtonProps> {
    public static readonly contextType = CalculatorControlContext;
    public readonly context: CalculatorControlContextValue;

    public get isDisabled(): boolean {
        return this.props.increase
            ? this.context.value >= this.context.max
            : this.context.value <= this.context.max;
    }

    public render(): JSX.Element {
        const { increase, ...props } = this.props;
        return (
            <button {...props} disabled={this.isDisabled} onClick={this.handleClick}>
                {this.props.children}
            </button>
        );
    }

    protected handleClick = () => {
        if (this.isDisabled) {
            return;
        }
        const delta = this.props.increase ? 1 : -1;

        this.context.onChange(this.context.value + delta * this.context.step);
    }
}
