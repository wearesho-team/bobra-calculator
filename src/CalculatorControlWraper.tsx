import * as React from "react";
import { CalculatorContext } from "./CalculatorContext";
import { CalculatorControlContext } from "./CalculatorControlContext";

export enum CalculatorControlTypes {
    term = "term",
    amount = "amount",
}

export interface CalculatorControlWrapperProps {
    type: CalculatorControlTypes,
}

export const CalculatorControlWrapper: React.FunctionComponent<CalculatorControlWrapperProps> = (props) => (
    <CalculatorContext.Consumer>
        {(value) => <CalculatorControlContext.Provider children={props.children} value={value[ props.type ]}/>}
    </CalculatorContext.Consumer>
);
