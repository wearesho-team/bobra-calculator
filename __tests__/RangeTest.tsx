import * as React from "react";
import { mount } from "enzyme";
import { spy } from "sinon";

import * as Calculator from "../src";

describe("<Range />", () => {
    it("have to match snapshot", () => {
        const wrapper = mount(
            <Calculator.ControlContext.Provider
                value={{ min: 500, max: 2000, step: 10, value: 1000, onChange: () => 1000 }}
            >
                <Calculator.Range />
            </Calculator.ControlContext.Provider>
        );
        expect(wrapper.html()).toMatchSnapshot();
    });

    it("have to trigger context onChange", () => {
        const handleChangeSpy = spy((n) => n);
        const wrapper = mount(
            <Calculator.ControlContext.Provider
                value={{ min: 500, max: 2000, step: 10, value: 1000, onChange: handleChangeSpy }}
            >
                <Calculator.Range />
            </Calculator.ControlContext.Provider>
        );
        wrapper.find("input").simulate("change", {
            target: {
                value: 1200,
            },
        });
        expect(handleChangeSpy.calledWith(1200)).toBeTruthy();
    });
});
