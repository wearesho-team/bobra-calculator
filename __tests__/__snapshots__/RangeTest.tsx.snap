// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Range /> have to match snapshot 1`] = `"<input type=\\"range\\" min=\\"500\\" max=\\"2000\\" step=\\"10\\" value=\\"1000\\">"`;
